PS-20 Code Completion for Emerald including submit() and get_job() functionality

andyjung-dev opened the pull request:
Quick commit: see [Trello Link](https://trello.com/c/GY72VojB/57-ps-17-code-completion-for-emerald-25)

andyjung-dev committed:
PS-20 Adding functionality for submit()

andyjung-dev committed:
PS-20 Adding functionality for submit()

andyjung-dev committed:
PS-20 autocommit

andyjung-dev committed:
PS-20 autocommit

hanburger97 commented on src/models/miner.model.js:
```diff
@@ -40,19 +63,142 @@ class MinerModel {
+             Reconstruct the block from data,check to see that it matches the sent block
+            */
+            
+            const blockTemplate = getCurrentBlockTemplate();
```
Nit pick 
```suggestion
            const blockTemplate = BlockTemplateService.getBlock()            
```

hanburger97 commented on src/models/miner.model.js:
```diff
@@ -40,19 +63,142 @@ class MinerModel {
         return new Promise((resolve, reject) => {
-            resolve('Boilerplate for Miner::submit');
+            //miner, job, blockTemplate, nonce, resultHash
+            const minerData = 
```
the `return new Promise()` was just a place holder, please dont use it like that, many of our Services, Api calls are already returning promises, let'st just chain them instead.

hanburger97 commented on src/models/miner.model.js:
```diff
@@ -40,19 +63,142 @@ class MinerModel {
+            var hashDiff = diff.div(bignum.fromBuffer(new Buffer(hashArray)));
+        
+            if (hashDiff.ge(blockTemplate.difficulty)){
+                moneroApi.submit(block, function(error, result){
```
Returns a promise instead of taking a callback
```suggestion
                return moneroApi.submit(block).then((...) => {...})
```

hanburger97 reviewed (changes requested):
Just a couple of changes on the way our services and api use promises.
Otherwise LGTM, make the small changes and 🚢  it

hanburger97 committed:
[PS14] Adding subscription route for Pickaxe (#8)

Co-Authored-By: Andy Jung <44798950+andyjung-dev@users.noreply.github.com>

andyjung-dev committed:
PS-20 Commit before rebase

andyjung-dev committed:
Will write up soon

andyjung-dev committed:
PS-20 spec

andyjung-dev committed:
Merge branch 'PS-20' of github.com:myriadeinc/emerald into PS-20

 PS-20

Refactored block submission logic

Currently it appears that a lot of the existing codebase for other pool servers utilize poorly documented features in communicating with the miner.

There are some functionalities such as `nextBlobWithChildNonce` that appear to only be in use for other cryptocurrencies.

Also, the reasoning behind `reserved_offset` also is poorly implemented so setting a fixed reserve offset of 8 bytes and allowing the miner/miner proxy to allocate resources from there should suffice.

The biggest todo is to find a nice way to document large changes such as this and to also have some repository where knowledge transfer can be available. Then we can do away with cluttered comments, but right now I am running low on sleep so this will have to wait.

andyjung-dev commented:
Check Slack for latest update

andyjung-dev committed:
Changes to Miner Model:
 - Moved submit() logic to block reference service
 - Moved getJob() logic to job helper service

hanburger97 commented on src/models/miner.model.js:
```diff
@@ -1,13 +1,41 @@
+ * Currently broken, we will need to fix
+ */
+const multiHash = require('mutli-hashing');
+const cryptoNight = mutliHash['cryptonight'];
```
Would this need to be changed to `randomX` and using a newer version of `multi-hashing` library??

hanburger97 commented on src/services/job.helper.service.js:
```diff
@@ -0,0 +1,55 @@
+     * @todo: add proper method
+     */
+    getFromId: (id) => {
+        blockTemplate = getLatestBlockTemplate();
```
So this method will check in the cache for a previously stored Job string and deserialize it into an Job object?? then we dont need `newJob` no?

hanburger97 commented on src/services/block.reference.service.js:
```diff
@@ -0,0 +1,80 @@
+    buildBlock: (minerData, job) => {
+        try{
+
+            var job = jobHelperService.getFromId(minerData.job_id);
```
nit: can we use `const` or `let` instead of `var`?

hanburger97 commented on src/services/block.reference.service.js:
```diff
@@ -0,0 +1,80 @@
+            var job = jobHelperService.getFromId(minerData.job_id);
+
+
+            var block = new Buffer(blockTemplate.buffer.length);
```
where does this `blockTemplate` comes from? is it suppose to be `job` bc that variable is not used in this function

hanburger97 commented on src/services/job.helper.service.js:
```diff
@@ -0,0 +1,55 @@
+                 */
+                target: difficulty.toString('hex'),
+        }
+        cache.put(newJob.id,newJob,"job");
```
nice!

hanburger97 replied:
nice!

hanburger97 commented on src/services/job.helper.service.js:
```diff
@@ -0,0 +1,55 @@
+        diffHex: getDiffHex(),
+        submissions: []
+        };
+        return cache.get(`job::${id}`).toJSON();
```
oh we should probably change the `cache.get` function to support namespacing instead using "job::job_id"

hanburger97 reviewed (commented):
Just a couple of small changes, we are almost there!! 😅

andyjung-dev committed:
PS-20 feat/refactor: Add new hashing utils and refactored part of submit() functionality
